### ISSUE TITLE
Chroma should instantiate meshes (Cube, Pyramid, Sphere) into all environments. Allowing environmental Enhancements such as Blender -> Chroma to work in all environments.

### ISSUE THREAD

**Issue by cal117** [Aeroluna/Heck#52]
Chroma should spawn objects in the environment that mappers can use for Blender environments such as the ones seen in Fragments, Somewhere Out There and Nothing Else + some other maps.

This would prevent modcharts that use these blender scripts from being limited to BTS Environment and the Gaga Environment.

The meshes that could be useful:

- Cube
- Triangle
- Sphere

**Comment by cal117** [comment]
A few people have requested this in the Beat Saber Mapping discord, thought I'd share it here.

**Comment by Fernthedev** [comment]
While I have suggested this in the past, I don't believe Chroma is required be behind this feat. You could easily make a mod that instantiates these objects before Chroma parses the environment and make it a map requirement. Personally I see it out of the scope of Chroma itself but that's just me.

**Comment by cal117** [comment]
Its another mod to maintain though - why bother for something so simple to implement?

**Comment by cal117** [comment]
Its literally:

`onsceneload... blah += ... if scene is a game ...`

```cs
Object.Instantiate(cubePrefab)
Object.Instantiate(pyramidPrefab)
Object.Instantiate(spherePrefab)
```

**Comment by Fernthedev** [comment]
Yeah, so make the other mod since it's so simple :)

**Comment by cal117** [comment]
Might as well then, it'll get me into PC modding

**Comment by Mawntee** [comment]
Fern, this should absolutely be a part of Chroma considering Chroma is the only thing that will be using it. 
Adding another weird requirement to the mix will confuse the crap out of the general player base. 

Creating a standard dark material 1x1 cube, then making it inactive shouldn't add much of a performance hit at all to maps that wouldn't be using it? 

I was also thinking having a random 1x1 light box that could be assigned through Light ID's would be incredible from a mapping perspective, but I'm not sure what that would do behind the scenes.

**Comment by Fatalution** [comment]
Seconding @Mawntee , this SHOULD NOT be a separate mod. Chroma and NE alone are too much 

for some players to handle. Adding another mod into the mix would be bad, especially considering that you would have to set it as requirement - otherwise people won't install it at all. And that would mean Chroma as requirement. No one checks the ? sign.

So, @Fernthedev , this should be a part of a Chroma if this is to exist. @cal117 maybe reopen this

**Comment by cal117** [comment]
Oh okay, will do

**Comment by Fernthedev** [comment]
Meh it's out of the scope of the mod itself, I don't really care much about the user if they can't figure out mods are required. But that's me, I don't handhold babies to play games. 
If it uses asset bundles you can count Quest out of the picture /shrug

**Comment by cal117** [comment]
It probably wont because meshes arent that hard to create programatically.

**Comment by Fatalution** [comment]
> Meh it's out of the scope of the mod itself, I don't really care much about the user if they can't figure out mods are required. But that's me, I don't handhold babies to play games. 

Well, do you map Chroma? Maybe you don't get what me and @Mawntee are coming from, but separating the mod into its own thing would make it so 5% of people AT MOST will play the map correctly. That's why Settings Setter was implemented. 

> If it uses asset bundles you can count Quest out of the picture /shrug

Cube, sphere and pyramid are base Unity assets that don't need asset bundles AFAIK.

**Comment by Fernthedev** [comment]
5% is a hyperbole at best. Settings setter doesn't exist on Quest, everything seems fine to me. As for the "downvote" spam mappers suffer blaming it on Quest, truth is most of them can't vote on BeatSaver unless they use PC in which case they **do** have a setting setter so the argument is nullified. 

To treat the user like a 5 year old or 90 year old is to disrespect them and confuse them further. They can make decisions, you just have to properly inform them. Don't give them a 20 page document nor a vague 5 word sentence. Setting setter does make that a better experience, but you could also just *not* rely on player specific settings to be the way you want (of course, that's not always possible but it's still the ideal way to go)

Separating functionality outside of Chroma, specifically creating new objects/meshes in the environment makes total sense to me because it could be extended with more functionality such as cube color, lighting, material etc. Sure maybe Chroma doesn't need all of that, but that's exactly why another mod would do best with it. Maybe it doesn't matter either. You don't see it that way because you look at this from a mapping perspective, not a functionality perspective. That always depends on who cares on what though.

As for cube, sphere, pyramid being base unity assets, I honestly don't know. Unity editor, perhaps. But being in the base game itself, especially Quest where unused assets/code are usually stripped makes it unlikely in my prediction. Maybe it does, maybe it doesn't. I haven't dealt with it so take it with a large bucket of salt. 

In the end, Aero is going to be the one approving or closing PRs for Chroma, and even if he does merge I can always just not port to Quest. Even then, as a separate mod it could still be PC specific. I know mappers generally don't care about Quest simply for not owning one, that's fine. If you want to use PC specific functionality, oh well I suppose. I'm not opposed to this functionality at all, just _how_ I'm going to deal with it myself since it's usually a "me" problem if it's part of Chroma.
That's pretty much it, no need to be dramatic about the "users" or whatever, just my stance on this specific feature.

tl;dr: asset bundles scare me

**Comment by cal117** [comment]
```cs
if (!scene.name.Contains("Environment")) return;

// Someone mind telling me a good way to get the Environment GO?
 GameObject environmentParent = GameObject.Find("")

 GameObject plane = GameObject.CreatePrimitive(PrimitiveType.Plane);

 // yeet
 plane.transform.position = new Vector3(0, -100000, 0);
 plane.transform.parent = // set parent to environment gameobject

 // todo repeat for others.

 GameObject cube = GameObject.CreatePrimitive(PrimitiveType.Cube);
 cube.transform.position = new Vector3(0, 0.5 f, 0);

 GameObject sphere = GameObject.CreatePrimitive(PrimitiveType.Sphere);
 sphere.transform.position = new Vector3(0, 1.5 f, 0);

 GameObject capsule = GameObject.CreatePrimitive(PrimitiveType.Capsule);
 capsule.transform.position = new Vector3(2, 1, 0);

 GameObject cylinder = GameObject.CreatePrimitive(PrimitiveType.Cylinder);
 cylinder.transform.position = new Vector3(-2, 1, 0);
```

**Comment by cal117** [comment]
These primatives do not have a scale by default, mapper will have to change the scale if they duplicate the objects or its useless - which 9/10 they will.

**Comment by Fernthedev** [comment]
They should probably have a reference/default scale so mappers **know** these objects are in the game. Besides that, seems good and hopefully it isn't stripped on Quest to make my life pain.
Good luck then

**Comment by cal117** [comment]
You can create the meshes manually on quest, check this out https://gist.github.com/prucha/866b9535d525adc984c4fe883e73a6c7
